### PR TITLE
Remote caching

### PR DESCRIPTION
**Pull request by regnat**
A first attempt at adding remote caching to funflow.

My first idea was to make the store itself abstract from the underlying filesystem, but since actions can directly access the store items as local directories it was much simpler to keep the store as-it-is and implement a remote cache on top of it (similar to the nix/bazel remote caches which come in addition to the local store).

At the moment the interface for the remote cache (defined in [Control.Funflow.RemoteCache](https://github.com/tweag/funflow/blob/934701c169d7322dbf812e47ee7f75a6fe1d7000/funflow/src/Control/Funflow/RemoteCache.hs)) is a simple typeclass providing two `push` and `pull` methods for saving and loading from the cache (though tha's definitely open for discussion because I'm not sure that's the best interface at all).
Whenever the executor tries to query a store path which isn't realized yet, it tries to pull it from the cache, and whenever it marks a store as complete, it pushes it to the cache.

### CODE REVIEW

**Comment by regnat** [comment]
Don't get too excited, it's really basic and probably not totally correct right now ;)

File: funflow/src/Control/Funflow/Exec/Simple.hs [hunk 1]
```diff
@@ -85,7 +87,8 @@ runFlowEx _ cfg store runWrapped confIdent flow input = do
           do
              liftIO $ BS.writeFile (toFilePath $ fp </> [relfile|out|])
                          . cacherStoreValue c $ res
-             _ <- CS.markComplete store chash
+             finalItem <- CS.markComplete store chash
+             _ <- Remote.push cacher chash (CS.itemPath store finalItem)
```
Comment (nc6) [inline funflow/src/Control/Funflow/Exec/Simple.hs]: Might be nice to have some logging here about the status of the remote push?

**Comment by nc6** [comment]
Looks like this is failing CI

**Comment by regnat** [comment]
Strange, the test pass fine on my machine (even when trying it on the same docker image as the CI).
I'll try to investigate this more

**Comment by YPares** [comment]
To avoid future conflicts, I merged https://github.com/tweag/funflow/pull/139 in this branch.

Some code could be factorized this way.

**Pull request merged**